gadget/mountedfilesystem.go

```diff
@@ -0,0 +1,184 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package gadget
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/strutil"
+)
+
+type MountedFilesystemWriter struct {
+	contentDir string
+	ps         *PositionedStructure
+}
+
+// NewMountedFilesystemWriter returns a writer capable of deploying provided
+// structure, with content of the structure stored in the given root directory.
+func NewMountedFilesystemWriter(contentDir string, ps *PositionedStructure) (*MountedFilesystemWriter, error) {
+	if ps == nil {
+		return nil, fmt.Errorf("internal error: *PositionedStructure is nil")
```

Comment:
The `*` can be understood as either pointer dereference or type name. Perhaps we can just tell which variable was nil?

```suggestion
		return nil, fmt.Errorf("internal error: ps is nil")
```

Reply:
`*PositionedStructure ..` matches Raw{Writer,Updater}. The error message returned there went through some discussion and ended up with this.